core/src/main/java/org/pentaho/platform/engine/core/system/objfac/OSGIObjectFactory.java

```diff
@@ -257,9 +258,9 @@ public void init( String configFile, Object context ) {
         @Override public int compare( ServiceReference o1, ServiceReference o2 ) {
```

Comment:
I'm not sure we need to implement a comparator since the `ServiceReference` class implements the method `compareTo()` which does exactly what we're doing here.

See: https://osgi.org/javadoc/r4v43/core/org/osgi/framework/ServiceReference.html#compareTo(java.lang.Object)

> If this ServiceReference and the specified ServiceReference have the same service id they are equal. This ServiceReference is less than the specified ServiceReference if it has a lower service ranking and greater if it has a higher service ranking. Otherwise, if this ServiceReference and the specified ServiceReference have the same service ranking, this ServiceReference is less than the specified ServiceReference if it has a higher service id and greater if it has a lower service id.

If we need to implement a comparator, the following code will be a better optimization.

```
Object ranking = this.getProperty(Constants.SERVICE_RANKING);
// if the property is not supplied or of incorrect type, use the default
int rank1 = ((ranking != null && ranking instanceof Integer) ? ((Integer) ranking).intValue() : 0);
ranking = other.getProperty(Constants.SERVICE_RANKING);
int rank2 = ((ranking != null && ranking instanceof Integer) ? ((Integer) ranking).intValue() : 0);

return rank1 - rank2;
```